### ISSUE TITLE
chef_solo provision wrong detection of chef existance for Windows guest

### ISSUE THREAD

**Issue by 111andre111** [hashicorp/vagrant#12346]
### Vagrant version
2.2.15

### Host operating system
OS Name : "Microsoft Windows 10 Pro 20H2"
OS Version: "10.0.19042 Build 19042"

### Guest operating system
OS Name : "Microsoft Windows Server 2019 Datacenter Evaluation"
OS Version : "10.0.17763 Build 17763"

### Vagrantfile

```ruby
# -*- mode: ruby -*-
# vi: set ft=ruby :
# All Vagrant configuration is done below. The "2" in Vagrant.configure
# configures the configuration version (we support older styles for
# backwards compatibility). Please don't change it unless you know what
# you're doing.
Vagrant.configure("2") do |config|
  config.vm.box = "StefanScherer/windows_2019"
  config.vm.communicator = "winrm"
  config.vm.network "forwarded_port", host: 2200, guest: 3389
  config.vm.provider "virtualbox" do |v|
    v.cpus = 2
    v.memory = 2048
  end
  config.winrm.timeout =   1800 # 30 minutes
  config.vm.boot_timeout = 1800 # 30 minutes
  config.winrm.max_tries = 600
  config.winrm.retry_delay = 5
  config.vm.provision "chef_solo" do |chef|
    # chef.cookbooks_path = ['C:\vagrant']
    chef.add_recipe "example-windowsserver2019"
    chef.arguments = "--chef-license accept"
    chef.run_list = [
      "recipe[chef-default]"
    ]
    # chef.install = false
  end
end
```

Please ensure the Vagrantfile provided is a minimal Vagrantfile which contains
only the required configuration to reproduce the behavior. Please note that if
your Vagrantfile contains an excess of configuration unrelated the the reported
issue, or is in a different format, we may be unable to assist with your issue.
Always start with a minimal Vagrantfile and include only the relevant configuration
to reproduce the reported behavior.

### Debug output
https://gist.github.com/111andre111/73d31bf50f0f7576b0d07a9d32323262

### Expected behavior
Using `chef_solo` provisioned `chef-solo` should be found to be installed, but it doesn't.

### Actual behavior
At this point in time, the chef_solo provisioned results in following error:
```
DEBUG winrmshell: [WinRM] remote shell created with shell_id: 91B40CC4-9705-4413-9E6A-B50A68FE59AB
DEBUG winrmshell: [WinRM] Command created for if ((&knife --version) -Match "Chef*"){ exit 0 } else { exit 1 }
if (!$?) { if($LASTEXITCODE) { exit $LASTEXITCODE } else { exit 1 } } with id: 60EF0B25-E775-430D-A7C4-56CFC8A77E68
DEBUG winrmshell: [WinRM] creating command_id: 60EF0B25-E775-430D-A7C4-56CFC8A77E68 on shell_id 91B40CC4-9705-4413-9E6A-B50A68FE59AB
DEBUG winrmshell: [WinRM] Waiting for output...
DEBUG winrmshell: [WinRM] Processing output
DEBUG winrmshell: [WinRM] cleaning up command_id: 60EF0B25-E775-430D-A7C4-56CFC8A77E68 on shell_id 91B40CC4-9705-4413-9E6A-B50A68FE59AB
DEBUG winrmshell: Output: #<WinRM::Output:0x0000020babd01bd8 @data=[{:stderr=>"The term 'knife' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.\nAt line:1 char:7\r\n+ if ((&knife --version) -Match \"Chef*\"){ exit 0 } else { exit 1 }\r\n+       ~~~~~\n    + CategoryInfo          : ObjectNotFound: (knife:String) [], CommandNotFoundException\n    + FullyQualifiedErrorId : CommandNotFoundException\n"}], @exitcode=1>
ERROR warden: Error occurred: Vagrant could not detect Chef on the guest! Even after Vagrant
attempted to install Chef, it could still not find Chef on the system.
Please make sure you are connected to the Internet and can access
Chef's package distribution servers. If you already have Chef
installed on this guest, you can disable the automatic Chef detection
by setting the 'install' option in the Chef configuration section of
your Vagrantfile:
    chef.install = false
 INFO warden: Beginning recovery process...
 INFO warden: Recovery complete.
ERROR warden: Error occurred: Vagrant could not detect Chef on the guest! Even after Vagrant
attempted to install Chef, it could still not find Chef on the system.
Please make sure you are connected to the Internet and can access
Chef's package distribution servers. If you already have Chef
installed on this guest, you can disable the automatic Chef detection
by setting the 'install' option in the Chef configuration section of
your Vagrantfile:
    chef.install = false
```
Now the problem is, that this provisioned downloads the chef infra client for getting the solo binaries:
`. { iwr -useb https://omnitruck.chef.io/install.ps1 } | iex; install -project 'chef' -channel 'stable'`
and it will download this msi: https://downloads.chef.io/tools/download/infra-client?download=https://packages.chef.io/files/stable/chef/17.0.242/windows/2019/chef-client-17.0.242-1-x64.msi&os=Windows%202019&version=17.0.242
installing chef from this location: `C:\Users\vagrant\AppData\Local\Temp\chef-client-17.0.242-1-x64.msi`
Now, if you inspect the MSI file, knife doesn't exist with this client and is even not needed for chef-solo.

So in my opinion the client does her the steps wrong:
https://github.com/hashicorp/vagrant/blob/5b501a3fb05ed0ab16cf10991b3df9d231edb5cf/plugins/provisioners/chef/provisioner/chef_solo.rb#L61
`          verify_binary(chef_binary_path("chef-solo"))`
https://github.com/hashicorp/vagrant/blob/5b501a3fb05ed0ab16cf10991b3df9d231edb5cf/test/unit/plugins/provisioners/chef/cap/windows/chef_installed_test.rb#L32
`      subject.chef_installed(machine, "chef_solo", version)`
https://github.com/hashicorp/vagrant/blob/5b501a3fb05ed0ab16cf10991b3df9d231edb5cf/plugins/provisioners/chef/cap/windows/chef_installed.rb
```
          def self.chef_installed(machine, product, version)
            if version != :latest
              command = 'if ((&knife --version) -Match "' + version.to_s + '"){ exit 0 } else { exit 1 }'
            else
              command = 'if ((&knife --version) -Match "Chef*"){ exit 0 } else { exit 1 }'
            end
            machine.communicate.test(command, sudo: true)
```
Now, can't we just add here an additional if section that tests if chef-solo is existing and execute `chef-solo --version` in order to check if everything is fine?

### References

- https://github.com/hashicorp/vagrant/issues/12337

**Comment by 111andre111** [comment]
Right now I discovered the same is not working on a Linux client host:
https://github.com/hashicorp/vagrant/blob/v2.2.16/plugins/provisioners/chef/cap/windows/chef_installed.rb
https://github.com/hashicorp/vagrant/blob/v2.2.16/plugins/provisioners/chef/cap/linux/chef_installed.rb
and probably the other OSes like debian, suse, FreeBSD etc.:
https://github.com/hashicorp/vagrant/tree/v2.2.16/plugins/provisioners/chef/cap

**Comment by ziyan-junaideen** [comment]
I think I am having the same issue in macOS Catalina. This worked before. I accidentally destroyed the Vagrant VM thinking it was another and that's when I saw this when I provisioned it.

```
==> default: Running provisioner: chef_solo...
The chef binary (either `chef-solo` or `chef-client`) was not found on
the VM and is required for chef provisioning. Please verify that chef
is installed and that the binary is available on the PATH.
jdeen@iMac:/Volumes/Dev/Work/GivingFire/Projects/givingfire|master⚡
⇒  vagrant provision
==> default: The node path '/Volumes/Dev/Work/GivingFire/Projects/givingfire-cookbook/nodes' doesn't exist. Ignoring...
==> default: Running provisioner: chef_solo...
    default: Installing Chef (latest)...
^C==> default: Waiting for cleanup before exiting...
^[[AVagrant could not detect Chef on the guest! Even after Vagrant
attempted to install Chef, it could still not find Chef on the system.
Please make sure you are connected to the Internet and can access
Chef's package distribution servers. If you already have Chef
installed on this guest, you can disable the automatic Chef detection
by setting the 'install' option in the Chef configuration section of
your Vagrantfile:

    chef.install = false
```

What I do is to run it the first time, then Ctrl + C, add `chef.install = false` and then run `vagrant provision`. Seems to be going fine (so far 🤭)

**Comment by indygwyn** [comment]
What happens if you set/change the `chef.product` to `chef-workstation`?  I think this problem is related to chef-provisioner defaulting to latest and latest since Chef 17 doesn't include knife.  I've been having similar issues trying to make the chef-apply provisioner work (I still have an issue there because the chef-apply provisioner doesn't support adding chef.arguments to accept the license, but I'm pretty sure if you change your product to chef-workstation or lock your chef.version to pre-17 you should be good.

**Comment by collinmcneese** [comment]
@111andre111  this should be fixed now in Vagrant `2.2.19`

![image](https://user-images.githubusercontent.com/8185808/140649571-661984fc-05bf-43f0-8fb8-5d5bcb2329bd.png)

**Comment by 111andre111** [comment]
@collinmcneese 
Yes, by now confirming that it's running. Thank you a lot for all the effort. So closing this issue.
